---
title: "Untitled"
---

```{r setup, include=FALSE}
# install.packages("genbankr", ask=F) # v1.0.0
library(genbankr)
# install.packages("igraph", ask=F) # v1.0.0
library(igraph)
# BiocManager::install(c("STRINGdb"), ask=F) # , version="3.8") #  
# see https://bioconductor.org/packages/devel/bioc/vignettes/STRINGdb/inst/doc/STRINGdb.pdf
#STRINGdb$methods()              # To list all the methods available.
#STRINGdb$help("get_graph")      # To visualize their documentation.
library(STRINGdb)     # activate the STRINGdb library # eg species_id=9606 is Homo sapiens 
#install.packages("VennDiagram", ask=F)
library(VennDiagram)
#install.packages("rentrez", ask=F)
library(rentrez)
#install.packages("tidyverse", ask=F)
library(tidyverse)
# install.packages("dplyr", ask=F)
library(dplyr)
# install.packages("stringr", ask=F)
library(stringr)
#BiocManager::install("GenomicRanges", ask=F)
library(GenomicRanges)
#install.packages("readr", ask=F)
library(readr)
# install.packages("curl", ask=F)
library(curl)
# install.packages("httr", ask=F)
library(httr)


####### define functions #########

#--- function1 --- function for plasmids 
getplasmidgenes <- function(GBAccession_ID){ #  function 
  plasmid <- readGenBank(GBAccession(GBAccession_ID), partial=TRUE)
  # cds(plasmid)$product # gene long names
  # cds(plasmid)$gene    # gene short names
  plasmid_genes <- unique(sort(tolower(as.vector(na.omit( cds(plasmid)$gene))))) 
  plasmid_genes <- gsub("[()]",  replacement="", plasmid_genes) # remove ( + )
  # unique protein-coding genes all uppercase sorted #  length(plasmid_genes)
  return(tolower(plasmid_genes)) } # end function 

#--- function2 --- get plasmid information by species name as input #modified
getchrom.plasmid <- function(chrom_name){ 
  dt <- subset(table_plasmid,Chrom == chrom_name)
  names(dt)[names(dt) == "Plasmid"] <- chrom_name
  dt <- subset(dt, select = -c(Chrom))
  dt } 

#--- function3 --- get intersection of a chromosome across all plasmids 
jaccard <- function(a) {  # a is dataframe (e.g.only data of E.coli K12)
  intersection <- as.integer(a$Interacting_genes)
  chrom.total.genes <- as.integer(a$Total_Genes)
  plasmid.genes <- as.integer(a$Genes)
  union = chrom.total.genes + plasmid.genes - intersection
  return (intersection/union) #value of intersection between chromosome and plasmids 
}                             # (e.g.K12 vs pEK499, K12 vs pEK204, K12 vs pEK516)

#--- function4 --- get plasmid metadata: (1) plasmid nickname, (2) species,
# (3) strain, (4) length, (5) number of genes
#table <- setDT(table, keep.rownames = TRUE)[] #convert rowname to the first column
getdata <- function(GBAccession_ID){
  plasmid <- readGenBank(GBAccession(GBAccession_ID), partial=TRUE)
  #read in information from plasmid accession number
  w <- elementMetadata(sources(plasmid))
  #get metadata(type,organism,mol_type,strain,db_xref,plasmid,loctype,temp_grouping_id)
  name <- w$plasmid #plasmid name
  species <- w$organism #species
  species <- gsub(" ",  replacement="_", species)
  strain <- w$strain #strain
  genes <- unique(sort(tolower(as.vector(na.omit(cds(plasmid)$gene))))) #remove redundant genes
  genes <- gsub("[()]",  replacement="", genes) 
  genes <- length(genes)    #number of unique plasmid genes
  info <- seqinfo(plasmid)  #get plasmid information
  df_plasmid <- as.data.frame(info)
  length <- df_plasmid$seqlengths #plasmid length
  metadata <- c(name,species,strain,genes,length) #combine all data
}
######## end set up functions #########

getwd()
setwd("~/Google Drive/TDA_Summer2021/RMD_FILES")
input <- data.frame(read.csv("stringInput.csv")) # species, proteins, String_ID
input$Species <- gsub("/",  replacement="-", input$Species)
input$interactions <- rep(0, length(input$Species))

#create an empty table
table_plasmid <- data.frame(Plasmid=character(), Chrom=character(), Genes=integer(), 
                             Interacting_genes=integer(), Interactions=integer(),
                             Total_Genes=integer(), Total_Interactions=integer())
#input id for species ("K12", "O157H7", "CFT073"...) (from input.csv)
#
accessID2 = c("EU935739", #  pEK499
             "EU935740", #  pEK204
             "EU935738", #  pEK516
             "NZ_HG941719.1" # pEC958 2
)  # end vector # check in https://www.ncbi.nlm.nih.gov/nuccore

table_plasmid <- data.frame(read.csv("plasmids_species_pEK204.csv")) # read in

for (k in 4411:4445){           # change k  1:4445
      
     if((k != 25)&&(k != 99)&&(k != 235)&&(k != 246)&&(k !=266)&&(k !=345)&&(k !=349)&&(k !=376) # 6
     &&(k !=403)&&(k !=407)&&(k !=416)&&(k !=419)&&(k !=485)&&(k !=529)&&(k !=546) # 7
     &&(k !=559)&&(k !=622)&&(k !=691)&&(k !=4377)&&(k !=4378)# 5 => 18 not work
     ){  #  if
       
      string_db <- STRINGdb$new(version="11", species=input$ID[k], # new STRINGdb object
                                score_threshold=400, input_directory="")
      # string_db$proteins$preferred_name contains the list of genes, eg "DR97_1"
      
      mapped   <- string_db$map(data.frame(gene_name = string_db$proteins$preferred_name),
                                'gene_name', removeUnmappedRows=T) 
      # mapped is a table with gene_name and STRING_id      # str(mapped) } } 
      links_all1 = string_db$get_interactions(string_db$mp(string_db$proteins$preferred_name)) 
      links_all <- links_all1[!duplicated(links_all1[,c('from','to')]),]
      # links_all is a table with the STRING_ids of the 
      # connected proteins (cols 1 & 2) & their score (col 3)
      # links_all does not contain duplicates (unlike links_all1)

      # Now create a table for all genes with STRING_ids and combined score
      from_vector <- c()			# empty vector for gene names rather than STRING_ids
      to_vector <- c()				# empty vector for gene names rather than STRING_ids  
      for (i in 1:length(links_all$from)){		# for each STRING_id, extract the gene name
        from_vector[i] <- subset(mapped, STRING_id == links_all$from[i])$gene_name
        to_vector[i] <- subset(mapped, STRING_id == links_all$to[i])$gene_name }
      links_all$from_ID <- tolower(from_vector)
      links_all$to_ID <- tolower(to_vector) 
      # str(links_all)  # now: from, to, combined_score, from_ID, to_ID

      write.csv(links_all, paste("OUTPUT2/", input$Species[k], "_chrom_data.csv", sep=""))
      input$interactions[k] <- length(links_all$combined_score)
      
   ######## get plasmid's gene interactions # select plasmid-associated interactions
    for (jj in 1:length(accessID2)){
      access_ID2 = accessID2[jj] # select plasmid 
      p1 <- getplasmidgenes(access_ID2) # str(p1) # 84 genes, single vector of names
      
      if(length(p1) >0){ # if genes, else ignore
      mapped_p <- data.frame() # mapped_pEK499$STRING_id # has from/to data
      for (m in 1:length(p1)){	# get plasmid genes in K12
        mapped_p <- rbind(mapped_p, data.frame(subset(links_all, (to_ID==p1[m])|(from_ID==p1[m]))))}
      dim(mapped_p) # 684 for K12 and pEK499
      write.csv(mapped_p, paste("OUTPUT2/", input$Species[k], "_", access_ID2, ".csv", sep=""))
     
      # add row to plasmid table
      table_plasmid[length(table_plasmid$Plasmid)+1,] <-
           c(access_ID2, input$Species[k],  length(p1), # number of plasmid genes
             length(unique(c(mapped_p$from_ID, mapped_p$to_ID))), length(mapped_p$from),
            length(mapped[,1]), input$interactions[k])
      str(table_plasmid)
      
     if(dim(mapped_p)[1] > 0){ # if genes found

      pdf(paste("OUTPUT2/", input$Species[k], "_", access_ID2, "_plasmid-chrom_subnetwork.pdf", sep=""),
          width=16, height=20)
      string_db$plot_network(string_db$mp(unique(c(mapped_p$from_ID, mapped_p$to_ID)))) # make plot 
      dev.off()    # make a plot of the network for fun
      
      pdf(paste("OUTPUT2/", input$Species[k], "_", access_ID2, "_intraplasmid_subnetwork.pdf", sep=""),
          width=16, height=20)
      string_db$plot_network(string_db$mp(p1)) # make plot 
      dev.off()    # make a plot of the network for fun
      
      #  plot histogram of score distribution
      pdf(paste("OUTPUT2/", input$Species[k], "_", access_ID2, "_chrom_histogram.pdf",sep=""),
          width=8, height=10)
      hist(mapped_p$combined_score, xlab="score", xlim=c(0,1000), ylab="#pairs",
           cex=2, breaks=90,   xaxp=c(0,1000,40), yaxp=c(0,4000,40) )
      grid(nx=10, ny=10, col="lightgray", lty="dotted", equilogs = TRUE)
      dev.off()         # output = example_data 
        } # end if matching genes
      
      } # end if plasmid genes found
    } # end for plasmids
  } # end for each chrom
} # end if k matches
str(table_plasmid) # check
write.csv( table_plasmid, "plasmids_species_pEK204_3.csv")
tail(table_plasmid)
k
input$Species[k]
# table_plasmid <- table_plasmid[-c(17564, 17565),]

for (j in 1:length(id_name)){  #e.g. Five plasmids information for six species
  print(getchrom.plasmid(id_name[j]))  } # end loop

#-----add jaccard index of each chrom and plasmid into dataframe
table_jaccard <- data.frame(Chrom=character()) #create a dataframe with empty column
#create dataframe with column number of length(plasmid) per chrom,
#change column name to plasmid acc no.
table_jaccard <- cbind(table_jaccard, 
          setNames(data.frame(matrix(ncol = length(table_plasmid$Plasmid)/length(unique(table_plasmid$Chrom)), #total no. of plasmid per chrom 
                                     nrow = 0)), 
  table_plasmid$Plasmid[1:(length(table_plasmid$Plasmid)/length(unique(table_plasmid$Chrom)))]))
#plasmid accession number
#bind 1st column with other columns with Plasmid ID
for (i in 1:length(chrom)) { #put in jaccard index
  table_jaccard[length(table_jaccard$Chrom)+1,] <-
  c(chrom[i], format(round(jaccard(subset(table_plasmid,Chrom == chrom[i])),3), nsmall=3))}

table_jaccard <- table_jaccard[, colSums(table_jaccard > 0.001) > 0]
write.csv(table_jaccard,"table_jaccard_species.csv") # write out to CSV

table_interaction <- c() #-----table of interaction
table_interaction <- data.frame(Chrom=character()) # dataframe with empty column
#create new columns with the length of plasmid per chrom,
#column name with plasmid acc no.
other.col <-  setNames(data.frame(matrix(ncol = length(table_plasmid$Plasmid)/length(unique(table_plasmid$Chrom)), #total no. of plasmid per chrom 
                                     nrow = 0)), 
table_plasmid$Plasmid[1:(length(table_plasmid$Plasmid)/length(unique(table_plasmid$Chrom)))])
#set col names with plasmid accession number
table_interaction <- cbind(table_interaction,other.col)
#bind  1st column with other columns 

for (i in 1:length(id_name)) { #add in interactions
  table_interaction[length(table_interaction$Chrom)+1,] <-
    c(id_name[i],  getchrom.plasmid(id_name[i])$Interactions)
  }
table_interaction <- table_interaction[, colSums(table_interaction != 0) > 0]
write.csv(table_interaction,"table_interaction_species.csv") # write out to CSV

#-----table of interacting genes
table_interacting_genes <- data.frame(Chrom=character()) # dataframe with an empty column
#create new columns with the length of plasmid per chrom,
#column name with plasmid acc no.
table_interacting_genes <- cbind(table_interacting_genes,other.col) 
# bind the first column with other columns with Plasmid ID

for (i in 1:length(id_name)) { #add in interacting genes
  table_interacting_genes[length(table_interacting_genes$Chrom)+1,] <-
    c(id_name[i],  getchrom.plasmid(id_name[i])$Interacting_genes) }
table_interacting_genes <-
   table_interacting_genes[, colSums(table_interacting_genes != 0) > 0]
write.csv(table_interacting_genes,"table_interacting_genes_species")

for (i in 1:length(accessID2)) { #show metadata of plasmids
  print(getdata(accessID2[i])) }

#---- table shows with plasmid metadata 
mt <- c("name","species","strain","genes","length")
table_metadata <- data.frame(Accession=character()) #create a dataframe with an empty column
m <- setNames(data.frame(matrix(ncol = length(mt), nrow = 0)),mt)
#create dataframe with column number of length(mt),
#change colnames with variables in mt
table_metadata <- cbind(table_metadata,m) #combine the first column with other variables

for (i in 1:length(accessID2)) { #get plasmids metadata
  table_metadata[length(table_metadata$Accession)+1,] <-
       c(accessID2[i], getdata(accessID2[i])) } #  end for
t(table_metadata)
write.csv( t(table_metadata), "plasmids_species_metadata.csv")
```