---
title: "Untitled"
---

```{r setup, include=FALSE}
# install.packages("genbankr", ask=F) # v1.0.0
library(genbankr)
# install.packages("igraph", ask=F) # v1.0.0
library(igraph)
# BiocManager::install(c("STRINGdb"), ask=F) # , version="3.8") #  
# see https://bioconductor.org/packages/devel/bioc/vignettes/STRINGdb/inst/doc/STRINGdb.pdf
#STRINGdb$methods()              # To list all the methods available.
#STRINGdb$help("get_graph")      # To visualize their documentation.
library(STRINGdb)     # activate the STRINGdb library # eg species_id=9606 is Homo sapiens 
#install.packages("VennDiagram", ask=F)
library(VennDiagram)
#install.packages("rentrez", ask=F)
library(rentrez)
#install.packages("tidyverse", ask=F)
library(tidyverse)
# install.packages("dplyr", ask=F)
library(dplyr)
# install.packages("stringr", ask=F)
library(stringr)
#BiocManager::install("GenomicRanges", ask=F)
library(GenomicRanges)
#install.packages("readr", ask=F)
library(readr)
# install.packages("curl", ask=F)
library(curl)
# install.packages("httr", ask=F)
library(httr)


####### define functions #########

#--- function1 --- function for plasmids 
getplasmidgenes <- function(GBAccession_ID){ #  function 
  plasmid <- readGenBank(GBAccession(GBAccession_ID), partial=TRUE)
  # cds(plasmid)$product # gene long names
  # cds(plasmid)$gene    # gene short names
  plasmid_genes <- unique(sort(tolower(as.vector(na.omit( cds(plasmid)$gene))))) 
  plasmid_genes <- gsub("[()]",  replacement="", plasmid_genes) # remove ( + )
  # unique protein-coding genes all uppercase sorted #  length(plasmid_genes)
  return(tolower(plasmid_genes)) } # end function 

#--- function2 --- get plasmid information by species name as input #modified
getchrom.plasmid <- function(chrom_name){ 
  dt <- subset(table_plasmid,Chrom == chrom_name)
  names(dt)[names(dt) == "Plasmid"] <- chrom_name
  dt <- subset(dt, select = -c(Chrom))
  dt } 

#--- function3 --- get intersection of a chromosome across all plasmids 
jaccard <- function(a) {  # a is dataframe (e.g.only data of E.coli K12)
  intersection <- as.integer(a$Interacting_genes)
  chrom.total.genes <- as.integer(a$Total_Genes)
  plasmid.genes <- as.integer(a$Genes)
  union = chrom.total.genes + plasmid.genes - intersection
  return (intersection/union) #value of intersection between chromosome and plasmids 
}                             # (e.g.K12 vs pEK499, K12 vs pEK204, K12 vs pEK516)

#--- function4 --- get plasmid metadata: (1) plasmid nickname, (2) species,
# (3) strain, (4) length, (5) number of genes
#table <- setDT(table, keep.rownames = TRUE)[] #convert rowname to the first column
getdata <- function(GBAccession_ID){
  plasmid <- readGenBank(GBAccession(GBAccession_ID), partial=TRUE)
  #read in information from plasmid accession number
  w <- elementMetadata(sources(plasmid))
  #get metadata(type,organism,mol_type,strain,db_xref,plasmid,loctype,temp_grouping_id)
  name <- w$plasmid #plasmid name
  species <- w$organism #species
  species <- gsub(" ",  replacement="_", species)
  strain <- w$strain #strain
  genes <- unique(sort(tolower(as.vector(na.omit(cds(plasmid)$gene))))) #remove redundant genes
  genes <- gsub("[()]",  replacement="", genes) 
  genes <- length(genes)    #number of unique plasmid genes
  info <- seqinfo(plasmid)  #get plasmid information
  df_plasmid <- as.data.frame(info)
  length <- df_plasmid$seqlengths #plasmid length
  metadata <- c(name,species,strain,genes,length) #combine all data
}
######## end set up functions #########

getwd()
setwd("~/Google Drive/TDA_Summer2021/RMD_FILES")
input <- data.frame(read.csv("stringInput.csv")) # species, proteins, String_ID
input$Species <- gsub("/",  replacement="-", input$Species)
input$interactions <- rep(0, length(input$Species))

chrom <- input$Species
length(chrom) # 4445
# Escherichia_coli_536_	818 = none?
# Escherichia_coli_ATCC_8739_	1153 = part fim plus part pap
# Escherichia_coli_BL21	1108  = fim plus part pap
# Escherichia_coli_CFT073_	355  = fim plus part pap
# Escherichia_coli_O157:H7_str._EDL933_	279 = fim
# Escherichia_coli_str._K-12_substr._MG1655_	1211 = fim
# Cellvibrio_sp._BR_	3041 = pil
# Pseudomonas_fluorescens_F113_	2566 = pil and fimA
# Pseudomonas_aeruginosa_	1 = pil 
# Aromatoleum_aromaticum_EbN1_	187 = pil 
# Klebsiella_oxytoca_	10 = fim
# Klebsiella_pneumoniae_	11 = fim
# Shigella_flexneri_2a_str._301_	351 = fim
# Methylibium_petroleiphilum_PM1_	1003 = pil and fimAB
# Rubrivivax_benzoatilyticus_JA2_ATCC_BAA-35_	2244 = none?
# Ralstonia_eutropha_H16_	852 = pil and fim 

o <- c(818, 1153, 1108, 355, 279, 1211, 3041, 2566, 1,
       187, 10, 11, 351, 1003, 2244, 852)
#create an empty table
table_plasmid <- data.frame(Plasmid=character(), Chrom=character(), Genes=integer(), 
                             Interacting_genes=integer(), Interactions=integer(),
                             Total_Genes=integer(), Total_Interactions=integer())
#input id for species ("K12", "O157H7", "CFT073"...) (from input.csv)

accessID2 = c(   "LT985289" )
#table_plasmid <- data.frame(read.csv("Table_S4_plasmids_species_pEK.csv")) # read in
#table_plasmid <- table_plasmid[,-c(1)]
#dim(table_plasmid) # 6153

for (k in 314:4445){           # change k  1:4445
#for (k in o){           # change k  1:4445
     if((k != 25)&&(k != 99)&&(k != 235)&&(k != 246)&&(k !=266)&&(k !=345)&&(k !=349)
        &&(k !=376)&&(k !=403)&&(k !=407)&&(k !=416)&&(k !=419)&&(k !=485)&&(k !=529)
        &&(k !=546)&&(k !=559)&&(k !=622)&&(k !=691)&&(k !=4377)&&(k !=4378)# 5 => 18 not work
         ){  #    
             print(k)
      string_db <- STRINGdb$new(version="11.5", species=input$ID[k], # new STRINGdb object
                                score_threshold=400, input_directory="")
      # string_db$proteins$preferred_name contains the list of genes, eg "DR97_1"
      mapped <- string_db$map(data.frame(gene_name = string_db$proteins$preferred_name),
                                'gene_name', removeUnmappedRows=T) 
      # mapped is a table with gene_name and STRING_id      # str(mapped) } } 
      links_all1 = string_db$get_interactions(string_db$mp(string_db$proteins$preferred_name)) 
      links_all <- links_all1[!duplicated(links_all1[,c('from','to')]),]
      # links_all is a table with the STRING_ids of the 
      # connected proteins (cols 1 & 2) & their score (col 3)
      # links_all does not contain duplicates (unlike links_all1)
      # Now create a table for all genes with STRING_ids and combined score
      from_vector <- c()			# empty vector for gene names rather than STRING_ids
      to_vector <- c()				# empty vector for gene names rather than STRING_ids  
      for (i in 1:length(links_all$from)){		# for each STRING_id, extract the gene name
        from_vector[i] <- subset(mapped, STRING_id == links_all$from[i])$gene_name
        to_vector[i] <- subset(mapped, STRING_id == links_all$to[i])$gene_name }
      links_all$from_ID <- tolower(from_vector)
      links_all$to_ID <- tolower(to_vector) 
      input$interactions[k] <- length(links_all$combined_score)
      
   ######## get plasmid's gene interactions # select plasmid-associated interactions
     access_ID2 = accessID2 # select plasmid 
     p1 <- getplasmidgenes(access_ID2) # str(p1) # 84 genes, single vector of names
    # access_ID2 = "synth"
     #   p1 <- c("fima","fimb","fimc","fimd","fime","fimf","fimg","fimh","fimh_2",
      #   "fimi","fima_2","fima_3","fima_4","fimc_2","fimd_3","fimd_4","fimd_5",
      #    "fimd_6","fimf_1","papa5","pilo2","pilt_1","pilv_1","pilv_2", # 42 genes
      #    "pap","papc","papd","pili","pilj","pilk","pill","pilm","piln","fims",
      #   "pilo","pilp","pilq","pilr","pils","pilt","pilu","pilv","pilvc","pilvd",
      #   "f7-2", "foog", "foob", "ecp_4533", "ecp_4536", "ecp_4538", "ecp_4539",
      #   "ecp_4540", "ecp_4541", "ecp_4532", "atgg", "atpb", "atpa", "atpd",
      #   "apth", "pape_2", "papa_2", "sfab", "papc_2", "papd_2", "atpf", "aptc",
      #   "apte", "paph_2", "papi", "papi_2", "papj", "papj_2", "papk", "fima_1") 
      
      mapped_p <- data.frame() # mapped_pEK499$STRING_id # has from/to data
      for (m in 1:length(p1)){	# get plasmid genes in K12
        mapped_p <- rbind(mapped_p, data.frame(subset(links_all,
                                          (to_ID==p1[m])|(from_ID==p1[m]))))}

      table_plasmid[length(table_plasmid$Plasmid)+1,] <- # add row to plasmid table
           c(access_ID2, input$Species[k],  length(p1), # number of plasmid genes
             length(unique(c(mapped_p$from_ID, mapped_p$to_ID))), length(mapped_p$from),
            length(mapped[,1]), input$interactions[k])

      if(dim(mapped_p)[1]>0){
       pdf(paste("INTRAPLASMID_FIMPAPIL/", input$Species[k], "_",
      #  pdf(paste("SYNTH/", input$Species[k], "_",
          access_ID2,"_intraplasmid_subnetwork.pdf", sep=""), width=16, height=20)
      string_db$plot_network(string_db$mp(p1)) # make plot 
      dev.off() }  # make a plot of the network for fun
  } # end for each chrom
} # end if k matches # 
table_plasmid2 <- distinct(table_plasmid)
table_plasmid2$Interacting_genes <- as.numeric(table_plasmid2$Interacting_genes)
table_plasmid2$Total_Genes <- as.numeric(table_plasmid2$Total_Genes)
table_plasmid2$Genes <- as.numeric(table_plasmid2$Genes)
table_plasmid2$Jaccard <- round(as.numeric(table_plasmid2$Interacting_genes)/(as.numeric(table_plasmid2$Total_Genes) + as.numeric(table_plasmid2$Genes)),3)
str(table_plasmid2) # check  
tail(table_plasmid2)
input$Species[k]
write.csv( table_plasmid2, "Table_Snew_plasmids_species_FIMPAPIL_2.csv")
k
# dim(table_plasmid2)[1]/4 # 1914 samples done
# table_plasmid <- table_plasmid[-c(6586),]
# str(table_plasmid) # check  

#---- table shows with plasmid metadata 
mt <- c("name","species","strain","genes","length")
table_metadata <- data.frame(Accession=character()) #create a dataframe with an empty column
m <- setNames(data.frame(matrix(ncol = length(mt), nrow = 0)),mt)
#create dataframe with column number of length(mt),
#change colnames with variables in mt
table_metadata <- cbind(table_metadata,m) #combine the first column with other variables

for (i in 1:length(accessID2)) { #get plasmids metadata
  table_metadata[length(table_metadata$Accession)+1,] <-
       c(accessID2[i], getdata(accessID2[i])) } #  end for
t(table_metadata)
write.csv( t(table_metadata), "plasmids_species_metadata.fimpappil.csv")
##########
```